pkg/scalers/kafka_x_scaler.go

```diff
@@ -0,0 +1,686 @@
+	if err != nil {
+		return []external_metrics.ExternalMetricValue{}, false, err
+	}
+	metric := GenerateMetricInMili(metricName, float64(totalLag))
+
+	return []external_metrics.ExternalMetricValue{metric}, totalLagWithPersistent > s.metadata.activationLagThreshold, nil
+}
+
+func (s *kafkaXScaler) TotalLag() (int64, int64, error) {
+	return s.getTotalLag()
+}
+
+// getTotalLag returns totalLag, totalLagWithPersistent, error
+// totalLag and totalLagWithPersistent are the summations of lag and lagWithPersistent returned by getLagForPartition function respectively.
+// totalLag maybe less than totalLagWithPersistent when excludePersistentLag is set to `true` due to some partitions deemed as having persistent lag
+func (s *kafkaXScaler) getTotalLag() (int64, int64, error) {
+	topicPartitions, err := s.getTopicPartitions()
+	if err != nil {
+		return 0, 0, err
+	}
+	s.logger.V(4).Info(fmt.Sprintf("Kafka scaler: Topic partitions %v", topicPartitions))
+
+	consumerOffsets, producerOffsets, err := s.getConsumerAndProducerOffsets(topicPartitions)
+	s.logger.V(4).Info(fmt.Sprintf("Kafka scaler: Consumer offsets %v, producer offsets %v", consumerOffsets, producerOffsets))
+	if err != nil {
+		return 0, 0, err
+	}
+
+	totalLag := int64(0)
+	totalLagWithPersistent := int64(0)
+	totalTopicPartitions := int64(0)
+
+	for topic, partitionsOffsets := range producerOffsets {
+		for partition := range partitionsOffsets {
+			lag, lagWithPersistent, err := s.getLagForPartition(topic, partition, consumerOffsets, producerOffsets)
+			if err != nil {
+				return 0, 0, err
+			}
+			totalLag += lag
+			totalLagWithPersistent += lagWithPersistent
+		}
+		totalTopicPartitions += (int64)(len(partitionsOffsets))
+	}
+	s.logger.V(1).Info(fmt.Sprintf("Kafka scaler: Providing metrics based on totalLag %v, topicPartitions %v, threshold %v", totalLag, topicPartitions, s.metadata.lagThreshold))
+
+	s.logger.V(1).Info(fmt.Sprintf("Kafka scaler: Consumer offsets %v, producer offsets %v", consumerOffsets, producerOffsets))
+
+	if !s.metadata.allowIdleConsumers {
+		// don't scale out beyond the number of topicPartitions
+		if (totalLag / s.metadata.lagThreshold) > totalTopicPartitions {
+			totalLag = totalTopicPartitions * s.metadata.lagThreshold
+		}
+	}
+	return totalLag, totalLagWithPersistent, nil
+}
+
+// getProducerOffsets returns the latest offsets for the given topic partitions
+func (s *kafkaXScaler) getProducerOffsets(topicPartitions map[string][]int) (map[string]map[int]int64, error) {
+	// Step 1: build one OffsetRequest
+	offsetRequest := make(map[string][]kafka.OffsetRequest)
+
+	for topic, partitions := range topicPartitions {
+		for _, partitionID := range partitions {
+			offsetRequest[topic] = append(offsetRequest[topic], kafka.FirstOffsetOf(partitionID), kafka.LastOffsetOf(partitionID))
+		}
+	}
+
+	// Step 2: send request
+	res, err := s.client.ListOffsets(context.Background(), &kafka.ListOffsetsRequest{
```

Comment:
same as above, we should propagate the context until here

Reply:
done